### PR TITLE
Benchmarks and performance enhancements

### PR DESCRIPTION
**Pull request by pioardi**
## PR Checklist

- [x] Please add a description of your changes.
- [x] We need your changes to come with unit tests in order to keep this project in quality and easy to maintain.
- [x] Please add a link to the open issue or task that this pull request will resolve.
- [ ] Add a note to the changelog to indicate the change.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

If yes please indicate it and migration info into the CHANGELOG.md file.

---

This PR contains an initial benchmark strategy and documentation and a performance enhancement

### CODE REVIEW

File: src/pools/abstract-pool.ts
Comment (jerome-benoit) [inline src/pools/abstract-pool.ts:174]: If you merge with master, that change will become unnecessary.

**Comment by Shinigami92** [comment]
I will look into this tomorrow
Good night :wave:

File: src/pools/abstract-pool.ts
Comment (pioardi) [reply src/pools/abstract-pool.ts:174]: I Will give It a try or of you want feel free ti change it

**Comment by pioardi** [comment]
> I will look into this tomorrow
> Good night 👋

Yes please bro if you can feel free to improve and clean up directly

File: src/pools/abstract-pool.ts
Comment (pioardi) [reply src/pools/abstract-pool.ts:174]: @jerome-benoit  I tried to merge with master and remove the unknown in favor of Response type but does not work.
Would you help me with those types please? I am too unfamiliar with TS

**Comment by pioardi** [comment]
I was also thinking the internalExecute changes on the Abstract Pool since that could be beneficial also for cluster pools.
I did it only for thread for my own simplicity

File: src/utility-types.ts
Comment (jerome-benoit) [inline src/utility-types.ts:44]: Could you please explain why you need that type definition>?

File: src/pools/abstract-pool.ts
Comment (pioardi) [reply src/pools/abstract-pool.ts]: Nevermind, I think I did it

File: benchmarks/benchmarks/package.json
Comment (jerome-benoit) [inline benchmarks/benchmarks/package.json:13]: Bump to the latest beta, and it will work better.

Comment (pioardi) [reply benchmarks/benchmarks/package.json:13]: I will do directly later, to test my performance enhancements I am now referring to a local path.
But thank you to catch anyway

File: benchmarks/benchmarks/fixed-poolifier.js
Comment (jerome-benoit) [inline benchmarks/benchmarks/fixed-poolifier.js:2]: You should not rely a the local build here but use the poolifier package in node_mobules. Otherwise the results are not replicable because not based on the very same code/version.

File: src/utility-types.ts
Comment (pioardi) [reply src/utility-types.ts:44]: We have a new data structure into the pool, the promiseMap ( maybe we can find a better name for that ).
Into this map the key is the messageId of each task submitted, and the value is given by the worker, the resolve function and the reject function.
When we receive a message from the worker we get a map entry and resolve/reject the promise based on the message.
Typescript asked me for types...and I did my best but I need help on that

File: benchmarks/benchmarks/fixed-poolifier.js
Comment (pioardi) [reply benchmarks/benchmarks/fixed-poolifier.js:2]: Yes I agree, as I said this is done to test the performance enhancement that is into this PR, shall be removed for sure and tested with another beta

**Comment by jerome-benoit** [comment]
I think you should: 

+ Remove other pool packages dev dependencies in the main package.json;
+ Make hyperfine a dev dependencies with scripts to use it; 
+ Merge and remove the previous pool benchmarking code.

**Comment by pioardi** [comment]
> I think you should:
> 
> * Remove other pool packages dev dependencies in the main package.json;
> * Merge and remove the previous pool benchmarking code.

We could remove the other packages and run an npm install --save-dev OTHER_TOOL when we need? 
Yes other benchmarking code will be almost removed :)

**Comment by jerome-benoit** [comment]
> > I think you should:
> > 
> > * Remove other pool packages dev dependencies in the main package.json;
> > * Merge and remove the previous pool benchmarking code.
> 
> We could remove the other packages and run an npm install --save-dev OTHER_TOOL when we need?

The main package.json does not need at all other pools package. Event the previous pools benchmarking code works once you do a `npm install` in benchmarks directory and run `node myBench.js` in it.

**Comment by pioardi** [comment]
> > > I think you should:
> > > 
> > > * Remove other pool packages dev dependencies in the main package.json;
> > > * Merge and remove the previous pool benchmarking code.
> > 
> > 
> > We could remove the other packages and run an npm install --save-dev OTHER_TOOL when we need?
> 
> The main package.json does not need at all other pools package. Event the previous pools benchmarking code works once you do a `npm install` in the benchmarks directory and run `node myBench.js` in it.

I am not sure I go it, in the main package.json we don't have other tools.
We have other tools only in the benchmarks package.json

**Comment by jerome-benoit** [comment]
> > > > I think you should:
> > > > 
> > > > * Remove other pool packages dev dependencies in the main package.json;
> > > > * Merge and remove the previous pool benchmarking code.
> > > 
> > > 
> > > We could remove the other packages and run an npm install --save-dev OTHER_TOOL when we need?
> > 
> > 
> > The main package.json does not need at all other pools package. Event the previous pools benchmarking code works once you do a `npm install` in the benchmarks directory and run `node myBench.js` in it.
> 
> I am not sure I go it, in the main package.json we don't have other tools.

Open and read the file and you see two of them ;)

**Comment by pioardi** [comment]
> package.json

OMG you are right

File: src/pools/thread/fixed.ts
Comment (jerome-benoit) [inline src/pools/thread/fixed.ts:119]: You can simplify it to: `value?.worker && value?.resolve && value?.reject`

Comment (pioardi) [reply src/pools/thread/fixed.ts]: Please check now I read your comment later but I marked worker, resolve and reject as mandatory on "JustTempValue"

Comment (pioardi) [reply src/pools/thread/fixed.ts]: > You've deleted too much ...
> The code to benchmark the poolifier pools must stay as well as worker choice algos. Your current work is meant to only replace benchmarking code with other pools:
> 
> * benchmarks/external-pools -> your work on other pools benchmarking
> * benchmarks -> stay as it is, just move the myBench.js and external directory to external-pools.

I did something like we now have two folders with a readme that describe the folders content.
We will go through that readme to add details on benchmarks vs other pools

**Comment by pioardi** [comment]
Next thing that I will do is clean up the bencharmks vs other tools to pass the npm run lint.
Need your help with the promiseMap field that should not be any on the declaration @jerome-benoit  @Shinigami92

**Comment by Shinigami92** [comment]
@pioardi I think I can help but cannot split myself into several pieces to work in parallel  
Trust me I'm human :wink:  
Yes I'm fast, but currently I'm working on the strategies branch

**Comment by pioardi** [comment]
> @pioardi I think I can help but cannot split myself into several pieces to work in parallel
> Trust me I'm human 😉
> Yes I'm fast, but currently I'm working on the strategies branch

Ahah that's fine take your time, in the meanwhile I am going to clean up a little more where I am able to :)

**Comment by pioardi** [comment]
@jerome-benoit  @Shinigami92  this PR is now more clean, just as summary:

- we need to remove the any and rename the JustTempValue type
- Decide if the new internalExecute can be now "promoted" to the AbstractPool ( I guess yes since it remove a lot of overhead having more and more listeners ). If yes we then can remove the maxTasks option.

**Comment by Shinigami92** [comment]
@pioardi so having a first look into this PR...

IMO we should split the PR into two:
1. reorganizing and building up the benchmark folder
2. add performance improvements

The first one should not contain changes within `src` folder

I think you will be happy some day when looking up into a clean PR instead of looking up for the one PR that does contain performance improvements but it was a benchmark PR and you didn't know anymore that it was in the benchmark PR... :shrug: that's just chaotic

---

Beside that, we could create a sub-project in the `benchmarks` folder and it has it's own `package.json`
A bit like a monorepo, but sadly I have zero knowledge how to set it up in a way to be controlled from the root project
So maybe it's just okay to when to benchmark, then you need to `cd benchmarks && npm run benchmark`

**Comment by pioardi** [comment]
> @pioardi so having a first look into this PR...
> 
> IMO we should split the PR into two:
> 
> 1. reorganizing and building up the benchmark folder
> 2. add performance improvements
> 
> The first one should not contain changes within `src` folder
> 
> I think you will be happy some day when looking up into a clean PR instead of looking up for the one PR that does contain performance improvements but it was a benchmark PR and you didn't know anymore that it was in the benchmark PR... 🤷 that's just chaotic
> 
> Beside that, we could create a sub-project in the `benchmarks` folder and it has it's own `package.json`
> A bit like a monorepo, but sadly I have zero knowledge how to set it up in a way to be controlled from the root project
> So maybe it's just okay to when to benchmark, then you need to `cd benchmarks && npm run benchmark`

So about the last point, we exactly did that, we created a sub-project with a package.json ( external-pools is a sub-project ).
To have 2 Pull request well, in general I would agree with you but you know changes are almost done now into this branch another couple of strokes and we have done

**Comment by pioardi** [comment]
I mean, performance enhancement is really into 1 file so is true that are 2 different things into one PR, but one is little

**Comment by Shinigami92** [comment]
> I mean, performance enhancement is really into 1 file so is true that are 2 different things into one PR, but one is little

This is just laziness but you are the boss
I need to understand what you added here with this performance improvement...

**Comment by pioardi** [comment]
> > I mean, performance enhancement is really into 1 file so is true that are 2 different things into one PR, but one is little
> 
> This is just laziness but you are the boss
> I need to understand what you added here with this performance improvement...

Yeah Sunday laziness lol
Anyway I mean I agree with your feedback but since we have already everything here why split things just for one file

File: src/pools/thread/fixed.ts
Comment (pioardi) [inline src/pools/thread/fixed.ts:36]: @Shinigami92  just to simplify you, what I did is almost into this JS doc.
When a task is execute we put a record into this map, and when we receive a message from the worker we got the record and resolve/reject the promise.
This is an enhancement because we now avoid to subscribe more and more listeners ( one per task ), and from there I was also thinking to "promote" this on the abstract pool so that cluster pool will take also benefits from this new mechanism

**Comment by Shinigami92** [comment]
> Yeah Sunday laziness lol
> Anyway I mean I agree with your feedback but since we have already everything here why split things just for one file

Because it's hard for me, your team-worker, to get into the benchmark vs source-code files in the [review tab](https://github.com/pioardi/poolifier/pull/209/files) and check 28 different files at once

File: src/utility-types.ts
Comment (Shinigami92) [inline src/utility-types.ts:48]: @jerome-benoit @pioardi We need a name for this :thinking:

File: src/utility-types.ts
Comment (Shinigami92) [inline src/utility-types.ts:43]: We need a description for whatever this is

File: src/utility-types.ts
Comment (Shinigami92) [inline src/utility-types.ts:57]: We should add JSDoc to this property

File: src/pools/thread/fixed.ts
Comment (Shinigami92) [inline src/pools/thread/fixed.ts:116]: Is this only available in thread pools? If so: why?

**Review by Shinigami92 — CHANGES_REQUESTED** [review]

File: src/utility-types.ts
Comment (pioardi) [reply src/utility-types.ts:48]: PromiseResponseWrapper ? Mmmm

File: src/pools/thread/fixed.ts
Comment (pioardi) [reply src/pools/thread/fixed.ts:116]: i added this only to threads for simplicity, if you and @jerome-benoit do not see drawbacks I would add It directly to the abstract pool

**Comment by Shinigami92** [comment]
Okay so if I understand it correctly the `benchmarks/internal` folder contains benchmarks for our internal testing and playing with e.g. algos
And the `benchmarks/versus-external-pools` is the _poolifier against the world_ folder that is for demoing and :mechanical_arm:

**Comment by pioardi** [comment]
> > Yeah Sunday laziness lol
> > Anyway I mean I agree with your feedback but since we have already everything here why split things just for one file
> 
> Because it's hard for me, your team-worker, to get into the benchmark vs source-code files in the [review tab](https://github.com/pioardi/poolifier/pull/209/files) and check 28 different files at once

Oh ok mmmmmm I amo Sorry for that, yiu can also Mark the files as viewed so should be more Easy...
Mmmm anyway you are right please let me know of you feel Is too hard for you

File: src/utility-types.ts
Comment (pioardi) [reply src/utility-types.ts:57]: To keep It simple I would descrive It as the worker that has the task assigned

File: src/utility-types.ts
Comment (pioardi) [reply src/utility-types.ts:43]: A wrapper containing a worker and two functions One to risolve and One to reject the promise.


Mmmm could be a good start

**Review by pioardi — COMMENTED** [review]
> Okay so if I understand it correctly the `benchmarks/internal` folder contains benchmarks for our internal testing and playing with e.g. algos
> And the `benchmarks/versus-external-pools` is the _poolifier against the world_ folder that is for demoing and 🦾

Yep correct

File: benchmarks/versus-external-pools/bench.sh
Comment (Shinigami92) [inline benchmarks/versus-external-pools/bench.sh:18]: I cannot perform this script cause I don't have `hyperfine` installed

File: benchmarks/versus-external-pools/package.json
Comment (Shinigami92) [inline benchmarks/versus-external-pools/package.json:9]: We could add some script here so it's much easier and more guided how and what to use

File: benchmarks/versus-external-pools/package.json
Comment (Shinigami92) [inline benchmarks/versus-external-pools/package.json:13]: I intentionally set all the direct pool dependencies to exact version matches, so it is clear which version was compared against each other

**Comment by Shinigami92** [comment]
> Oh ok mmmmmm I amo Sorry for that, yiu can also Mark the files as viewed so should be more Easy...
> Mmmm anyway you are right please let me know of you feel Is too hard for you

It's fine for now, but keep this is mind for later PRs

File: src/pools/thread/fixed.ts
Comment (Shinigami92) [reply src/pools/thread/fixed.ts:120]: I cannot say too much to this because I'm not so deep in the technical aspects but more into the quality etc. ^^'
So @jerome-benoit needs to look into this

**Comment by Shinigami92** [comment]
Okay I think I'm out for today
I need some rest and play games and chill a bit
See ya tomorrow :wave:

File: benchmarks/versus-external-pools/package.json
Comment (pioardi) [reply benchmarks/versus-external-pools/package.json:9]: I would add all the instructions to the readme, I Will do that and from thrre se can evaluate of we Need also a script

File: benchmarks/versus-external-pools/package.json
Comment (pioardi) [reply benchmarks/versus-external-pools/package.json:13]: It Is ok

Comment (pioardi) [reply benchmarks/versus-external-pools/package.json:13]: For now

File: src/pools/thread/fixed.ts
Comment (pioardi) [reply src/pools/thread/fixed.ts:120]: @jerome-benoit I would try ti add It to the abstract pool of you agree

File: benchmarks/versus-external-pools/bench.sh
Comment (jerome-benoit) [reply benchmarks/versus-external-pools/bench.sh:18]: Unless you are on Windows, that's pretty easy to install: https://github.com/sharkdp/hyperfine

File: src/pools/thread/fixed.ts
Comment (jerome-benoit) [inline src/pools/thread/fixed.ts:46]: Why is it only added to the fixed pool class? 
It's a global performance optimisation, thus should be added to the abstract class.

File: src/pools/thread/fixed.ts
Comment (jerome-benoit) [inline src/pools/thread/fixed.ts:110]: Same remark here.

File: src/pools/thread/fixed.ts
Comment (jerome-benoit) [reply src/pools/thread/fixed.ts:120]: I fully agree. It's a global performance optimisation.

File: src/pools/thread/fixed.ts
Comment (pioardi) [reply src/pools/thread/fixed.ts:46]: Yes I was asking that, I Will add It to the cluster pools Aldo ;)

Comment (jerome-benoit) [reply src/pools/thread/fixed.ts:46]: If you add it to the abstract class, it's available for all pool/worker type.

Comment (pioardi) [reply src/pools/thread/fixed.ts:46]: Yes correct

File: benchmarks/versus-external-pools/bench.sh
Comment (jerome-benoit) [reply benchmarks/versus-external-pools/bench.sh:18]: Errata: it's also easily installable on windows with chocolatey.

File: src/pools/thread/fixed.ts
Comment (pioardi) [reply src/pools/thread/fixed.ts]: Done

File: benchmarks/versus-external-pools/bench.sh
Comment (pioardi) [reply benchmarks/versus-external-pools/bench.sh:18]: Added documentation on how to run bench

**Comment by pioardi** [comment]
Ok so guys, @Shinigami92  ( thank you thank you bro ) fixed the any type on the promiseMap and improved the general shape.
I added the performance enhancement to the abstract pool so that also the cluster pool will inherit this enhancement.

We just need to increase the coverage and I guess we have almost done

**Comment by pioardi** [comment]
Now this PR is almost ready to go guys, I will hold on to have your final review

File: benchmarks/README.md
Comment (Shinigami92) [inline benchmarks/README.md:15]: I don't understand this sentence

File: src/pools/cluster/fixed.ts
Comment (Shinigami92) [inline src/pools/cluster/fixed.ts:83]: Do we never remove the listener?

**Review by Shinigami92 — CHANGES_REQUESTED** [review]

**Comment by Shinigami92** [comment]
@pioardi 

> I also removed the unregisterWorkerMessageListener since is no more needed now.

Why isn't it needed anymore? Could you explain?
I don't really know if this could have bad side effects and even if the worker was removed from our pool, it never gets garbage collected because it has a listener attached :thinking:

File: src/pools/cluster/fixed.ts
Comment (pioardi) [reply src/pools/cluster/fixed.ts:83]: We do not need to remove the listener, we now have 1 listener for all the messages and all the workers.

File: benchmarks/README.md
Comment (pioardi) [reply benchmarks/README.md:15]: I mean "we will add more details"

File: src/pools/cluster/fixed.ts
Comment (pioardi) [reply src/pools/cluster/fixed.ts:83]: In other words, is one global listener for any message from any worker

Comment (Shinigami92) [reply src/pools/cluster/fixed.ts:83]: Ah, but we don't want to clean up the `listener` but the reference/attachment to the worker

File: benchmarks/README.md
Comment (Shinigami92) [reply benchmarks/README.md:15]:
```suggestion
We will add more details on each function that we benchmark.
```

**Comment by Shinigami92** [comment]
My working day begins, so I'm out for a while :wave:

File: src/pools/cluster/fixed.ts
Comment (pioardi) [reply src/pools/cluster/fixed.ts:83]: I think is not mandatory when a worker is destroyed the listener instance is no more attached.
To be more accurate we have one listener per worker, so when the worker got deleted bring with it also all the listeners ( i.e. we do not remove attached errorHandlers,onlinHandlers, and so on )

Comment (pioardi) [reply src/pools/cluster/fixed.ts:83]: @jerome-benoit  to share thoughts on that

Comment (jerome-benoit) [reply src/pools/cluster/fixed.ts:83]: If worker removal is properly implemented in node.js for cluster and worker-threads, that's indeed should not be necessary to do the cleanup in poolifier.

Comment (pioardi) [reply src/pools/cluster/fixed.ts:83]: Yes correct, so for now I guess we are ok with this

Comment (pioardi) [reply src/pools/cluster/fixed.ts:83]: Please @jerome-benoit  let me know if you have any other comment

**Review by jerome-benoit — APPROVED** [review]
Could you please update the PR description before the merge?

**Comment by pioardi** [comment]
> Could you please update the PR description before the merge?

I am going to, thanks jerome and shini

**Comment by pioardi** [comment]
I am going to increase our coverage as fast as I need and then we will release another beta ( or maybe an alpha at this point ? )

**Comment by Shinigami92** [comment]
> I am going to increase our coverage as fast as I need and then we will release another beta ( or maybe an alpha at this point ? )

The github action workflow is currently written in a way so it only supports beta or not beta

https://github.com/pioardi/poolifier/blob/be0676b3936d75f22ce55b0f71a1fb03d008a01c/.github/workflows/npmpublish.yml#L57-L67

**Comment by pioardi** [comment]
> > I am going to increase our coverage as fast as I need and then we will release another beta ( or maybe an alpha at this point ? )
> 
> The github action workflow is currently written in a way so it only supports beta or not beta
> 
> https://github.com/pioardi/poolifier/blob/be0676b3936d75f22ce55b0f71a1fb03d008a01c/.github/workflows/npmpublish.yml#L57-L67

Ok then we will go with another beta, not an issue bro

**Pull request merged**
